server/src/internalClusterTest/java/org/opensearch/cluster/routing/WeightedRoutingIT.java

```diff
@@ -857,4 +861,72 @@ public void testReadWriteWeightedRoutingMetadataOnNodeRestart() throws Exception
         );
 
     }
+
+    /**
+     * https://github.com/opensearch-project/OpenSearch/issues/18817
+     * For regression in custom string query preference with awareness attributes enabled.
+     * We expect preference will consistently route to the same shard replica. However, when awareness attributes
+     * are configured this does not hold.
+     */
+    public void testCustomPreferenceShardIdCombination() {
+        // Configure cluster with awareness attributes
+        Settings commonSettings = Settings.builder()
+            .put("cluster.routing.allocation.awareness.attributes", "rack")
+            .put("cluster.routing.allocation.awareness.force.rack.values", "rack1,rack2")
+            .put("cluster.routing.use_adaptive_replica_selection", false)
+            .put("cluster.search.ignore_awareness_attributes", false)
+            .build();
+
+        // Start cluster
+        internalCluster().startClusterManagerOnlyNode(commonSettings);
+        internalCluster().startDataOnlyNodes(2, Settings.builder().put(commonSettings).put("node.attr.rack", "rack1").build());
+        internalCluster().startDataOnlyNodes(2, Settings.builder().put(commonSettings).put("node.attr.rack", "rack2").build());
+
+        ensureStableCluster(5);
+        ensureGreen();
+
+        // Create index with specific shard configuration
+        assertAcked(
+            prepareCreate("test_index").setSettings(
+                Settings.builder().put("index.number_of_shards", 6).put("index.number_of_replicas", 1).build()
+            )
+        );
+
+        ensureGreen("test_index");
+
+        // Index test documents
+        for (int i = 0; i < 30; i++) {
+            client().prepareIndex("test_index").setId(String.valueOf(i)).setSource("field", "value" + i).get();
+        }
+        refresh("test_index");
```

Comment:
Maybe prefer `refreshAndWaitForReplication("test_index")` so that this test will work with segment replication?